File: behavioral/chain_of_responsibility/http_request_middleware/go/http_request_middleware/http_request_middleware_test.go
package httprequestmiddleware

import (
	"errors"
	"io"
	"os"
	"strings"
	"testing"
)

// Helper function to capture stdout
func captureOutput(f func()) string {
	old := os.Stdout // keep backup of the real stdout
	r, w, _ := os.Pipe()
	os.Stdout = w

	f() // Execute the function whose output we want to capture

	w.Close()
	os.Stdout = old // restoring the real stdout
	out, _ := io.ReadAll(r)
	return string(out)
}

// Test suite setup
func setupTestChain() Middleware {
	logger := &LoggingMiddleware{}
	authenticator := &AuthenticationMiddleware{}
	authorizer := &AuthorizationMiddleware{}
	finalHandler := &FinalHandler{}

	// Build chain: Log -> Auth -> Authorize -> Final
	logger.SetNext(authenticator).SetNext(authorizer).SetNext(finalHandler)
	return logger // Return the entry point
}

func TestMiddlewareChain(t *testing.T) {

	t.Run("Successful Request Admin", func(t *testing.T) {
		chain := setupTestChain()
		req := MakeRequest("/admin", "valid_token", "admin")

		var resp *Response
		var err error
		output := captureOutput(func() {
			resp, err = chain.Handle(req)
		})

		if err != nil {
			t.Errorf("Expected no error, but got: %v", err)
		}
		if resp == nil {
			t.Fatal("Expected a response, but got nil")
		}
		if resp.StatusCode != 200 {
			t.Errorf("Expected status code 200, but got %d", resp.StatusCode)
		}
		if !strings.Contains(resp.Body, "Successfully processed") {
			t.Errorf("Response body mismatch: %s", resp.Body)
		}
		if !strings.Contains(resp.Body, "user 789") {
			t.Errorf("Response body does not contain correct user ID: %s", resp.Body)
		}

		// Check logs
		if !strings.Contains(output, "[Log] Received request") { t.Error("Missing Log output") }
		if !strings.Contains(output, "[Auth] Authentication successful") { t.Error("Missing Auth success output") }
		if !strings.Contains(output, "[ACL] Authorization successful") { t.Error("Missing ACL success output") }
		if !strings.Contains(output, "[Final] Processing request") { t.Error("Missing Final processing output") }
	})

	t.Run("Authentication Failure", func(t *testing.T) {
		chain := setupTestChain()
		req := MakeRequest("/secure", "invalid_token", "")

		var resp *Response
		var err error
		output := captureOutput(func() {
			resp, err = chain.Handle(req)
		})

		if !errors.Is(err, ErrAuthenticationFailed) {
			t.Errorf("Expected ErrAuthenticationFailed, but got: %v", err)
		}
		if resp == nil {
			t.Fatal("Expected a response, but got nil")
		}
		if resp.StatusCode != 401 {
			t.Errorf("Expected status code 401, but got %d", resp.StatusCode)
		}
		if resp.Body != "Unauthorized" {
			t.Errorf("Expected body 'Unauthorized', got '%s'", resp.Body)
		}

		// Check logs show stop at Auth
		if !strings.Contains(output, "[Log] Received request") { t.Error("Missing Log output") }
		if !strings.Contains(output, "[Auth] Authentication failed") { t.Error("Missing Auth failure output") }
		if strings.Contains(output, "[ACL] Checking authorization") { t.Error("ACL should not have been checked") }
		if strings.Contains(output, "[Final] Processing request") { t.Error("Final Handler should not have been reached") }
	})

	t.Run("Authorization Failure", func(t *testing.T) {
		chain := setupTestChain()
		req := MakeRequest("/admin/super", "valid_token", "superadmin")

		var resp *Response
		var err error
		output := captureOutput(func() {
			resp, err = chain.Handle(req)
		})

		if !errors.Is(err, ErrAuthorizationFailed) {
			t.Errorf("Expected ErrAuthorizationFailed, but got: %v", err)
		}
		if resp == nil {
			t.Fatal("Expected a response, but got nil")
		}
		if resp.StatusCode != 403 {
			t.Errorf("Expected status code 403, but got %d", resp.StatusCode)
		}
		if resp.Body != "Forbidden" {
			t.Errorf("Expected body 'Forbidden', got '%s'", resp.Body)
		}

		// Check logs show stop at ACL
		if !strings.Contains(output, "[Log] Received request") { t.Error("Missing Log output") }
		if !strings.Contains(output, "[Auth] Authentication successful") { t.Error("Missing Auth success output") }
		if !strings.Contains(output, "[ACL] Authorization failed") { t.Error("Missing ACL failure output") }
		if strings.Contains(output, "[Final] Processing request") { t.Error("Final Handler should not have been reached") }
	})

	t.Run("Successful Request No Role Required", func(t *testing.T) {
		chain := setupTestChain()
		req := MakeRequest("/public", "valid_token", "") // No role required

		var resp *Response
		var err error
		output := captureOutput(func() {
			resp, err = chain.Handle(req)
		})


		if err != nil {
			t.Errorf("Expected no error, but got: %v", err)
		}
		if resp == nil {
			t.Fatal("Expected a response, but got nil")
		}
		if resp.StatusCode != 200 {
			t.Errorf("Expected status code 200, but got %d", resp.StatusCode)
		}
		if !strings.Contains(resp.Body, "Successfully processed") {
			t.Errorf("Response body mismatch: %s", resp.Body)
		}

		// Check logs
		if !strings.Contains(output, "[Log] Received request") { t.Error("Missing Log output") }
		if !strings.Contains(output, "[Auth] Authentication successful") { t.Error("Missing Auth success output") }
        if !strings.Contains(output, "[ACL] No specific role required. Proceeding.") { t.Error("Missing ACL no role required output") }
		if !strings.Contains(output, "[Final] Processing request") { t.Error("Missing Final processing output") }
	})
}

File: behavioral/chain_of_responsibility/http_request_middleware/typescript/http_request_middleware.test.ts
import {
  LoggingMiddleware,
  AuthenticationMiddleware,
  AuthorizationMiddleware,
  FinalHandler,
  makeHttpRequest,
  Middleware,
  HttpResponse,
} from "./http_request_middleware";

// Mock console.log to prevent test output clutter and allow assertions
let consoleOutput: string[] = [];
const originalConsoleLog = console.log;
const mockedConsoleLog = (output: string) => {
  consoleOutput.push(output);
};

describe("HttpRequest Middleware Chain", () => {
  let chain: Middleware;
  let logger: LoggingMiddleware;
  let authenticator: AuthenticationMiddleware;
  let authorizer: AuthorizationMiddleware;
  let finalHandler: FinalHandler;

  beforeEach(() => {
    // Reset console log mock before each test
    console.log = mockedConsoleLog;
    consoleOutput = [];

    // Setup the chain
    logger = new LoggingMiddleware();
    authenticator = new AuthenticationMiddleware();
    authorizer = new AuthorizationMiddleware();
    finalHandler = new FinalHandler();

    // Properly set up the chain
    logger.setNext(authenticator);
    authenticator.setNext(authorizer);
    authorizer.setNext(finalHandler);

    // Start with the logger
    chain = logger;
  });

  afterEach(() => {
    // Restore original console.log after each test
    console.log = originalConsoleLog;
  });

  test("should process a successful request for admin role", async () => {
    const request = makeHttpRequest("/admin", "valid_token", "admin");
    const response = await chain.handle(request);

    expect(response.statusCode).toBe(200);
    expect(response.body).toContain("Successfully processed request");
    expect(response.body).toContain("user 456"); // Check user info propagated

    // Check logs to ensure all middleware ran
    const logString = consoleOutput.join("\n");
    expect(logString).toContain("[Log] Received request");
    expect(logString).toContain("[Auth] Authentication successful");
    expect(logString).toContain("[ACL] Authorization successful");
    expect(logString).toContain("[Final] Processing request");
  });

  test("should reject request with invalid authentication", async () => {
    const request = makeHttpRequest("/secure", "invalid_token");
    const response = await chain.handle(request);

    expect(response.statusCode).toBe(401);
    expect(response.body).toBe("Unauthorized");

    // Check logs to ensure chain stopped at Auth
    const logString = consoleOutput.join("\n");
    expect(logString).toContain("[Log] Received request");
    expect(logString).toContain(
      "[Auth] Authentication failed. Aborting request."
    );
    expect(logString).not.toContain("[ACL] Checking authorization");
    expect(logString).not.toContain("[Final] Processing request");
  });

  test("should reject request with insufficient authorization", async () => {
    const request = makeHttpRequest(
      "/admin/super",
      "valid_token",
      "superadmin"
    );
    const response = await chain.handle(request);

    expect(response.statusCode).toBe(403);
    expect(response.body).toBe("Forbidden");

    // Check logs to ensure chain stopped at ACL
    const logString = consoleOutput.join("\n");
    expect(logString).toContain("[Log] Received request");
    expect(logString).toContain("[Auth] Authentication successful");
    expect(logString).toContain("[ACL] Authorization failed.");
    expect(logString).not.toContain("[Final] Processing request");
  });

  test("should process successfully when no specific role is required", async () => {
    const request = makeHttpRequest("/public", "valid_token");
    const response = await chain.handle(request);

    expect(response.statusCode).toBe(200);
    expect(response.body).toContain("Successfully processed request");

    // Check logs to ensure ACL passed correctly
    const logString = consoleOutput.join("\n");
    expect(logString).toContain("[Log] Received request");
    expect(logString).toContain("[Auth] Authentication successful");
    expect(logString).toContain("[ACL] No specific role required. Proceeding.");
    expect(logString).toContain("[Final] Processing request");
  });

  test("FinalHandler should not allow setting next", () => {
    const final = new FinalHandler();
    const another = new LoggingMiddleware();
    expect(() => final.setNext(another)).toThrow(
      "FinalHandler cannot have a next handler."
    );
  });
});
